common/tasks/interleaved_weighted_round_robin.go

```diff
@@ -202,6 +211,45 @@ func (s *InterleavedWeightedRoundRobinScheduler[T, K]) eventLoop() {
 	}
 }
 
+func (s *InterleavedWeightedRoundRobinScheduler[T, K]) cleanupLoop() {
+	defer s.shutdownWG.Done()
+	if s.options.InactiveChannelDeletionDelay == nil {
+		return
+	}
+	ch, _ := s.ts.NewTimer(s.options.InactiveChannelDeletionDelay())
+	for {
+		select {
+		case <-ch:
+			s.doCleanup()
+			ch, _ = s.ts.NewTimer(s.options.InactiveChannelDeletionDelay())
+		case <-s.shutdownChan:
+			return
+		}
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler[T, K]) doCleanup() {
+	s.Lock()
+	defer s.Unlock()
+	var keysToDelete []K
+	cleanupDelay := s.options.InactiveChannelDeletionDelay()
+	now := s.ts.Now()
+	for k, weightedChan := range s.weightedChannels {
+		if now.Sub(weightedChan.LastActiveTime()) > cleanupDelay && len(weightedChan.Chan()) == 0 {
```

Comment:
hmmm there might be a race condition here? doCleanup might decided to clean up this channel but (Try)Submit could already have a reference to the channel and about to push a task into it? 

We probably need a lock on the weightedChan, have it locked when returning the weightedChan from getOrCreateTaskChannel. Also return a releaseFn.

Reply:
Right! Thanks for catching that! I made the change to fix it. I used the existing lock. The read lock is released only after the task is written to the channel.

Reply:
hmm I don't think we should reuse the existing lock, otherwise Submit() on a full channel will cause the entire scheduler to stuck.

One way I can think of that doesn't require any change to WeightedChan is to track outstanding references of task channels and block the cleanup logic if needed. Of course, we need to notify those outstanding channel holders. 
Basically, in clean up, first s.Lock() to prevent any new Submit/TrySubmit from getting a reference to task channels, then send a notification to interrupt existing ones (e.g. close a channel), wait until all outstanding channel references are drained, then clean up channels. 
Submit() and TrySubmit() can simply retry if get interrupted by the notification. Upon retry, their logic will get blocked as the clean up logic already locked the scheduler.

Reply:
If we can get the channel and insert a message without releasing the lock, we can fix the race. I changed getOrCreateTaskChannel() to sendToChannel() to do that. Let me know what you think about this approach..

Reply:
My main concern is that a blocking Submit() call can prevent the entire scheduler from serving any new requests, because it holds the lock. 
Yeah I understand today there's no Submit() usage on this iwrr scheduler, but even for non-blocking Submit, the change essentially serializes all Submit/TrySubmit requests, and I am concerned this may have a performance hit. If there's a way to keep the original parallel behavior I think we should go for that.

Reply:
Ah that makes sense. 
But Submit() will only take RWLock() so it allows other Submit() and TrySubmit() to happen in parallel. There is on case when we call sendAndNotify(blocking method) with write lock. That is when the queue is first created. I have converted that to read lock now. It looks like there can be a race condition there. But I guess it is safe to assume that cleaner will not have deleted the channel that was created just now.

Now I think using a reference count in channel is a cleaner way to do this. And only cleanup if reference count is 0. Let me rewrite it that way.

Reply:
Sorry for the back and forth. Now I have simplified it using an atomic reference counter in WeightedChannel.
I have also added a test that is reliabily catching this race if it happens.

Reply:
Thank you! This looks great!